text/0000-more-api-documentation-conventions.md

```diff
@@ -0,0 +1,730 @@
+
+For the most part, the RFC proposes guidelines that are already followed today,
+but it tries to motivate and clarify them.
+
+# Detailed design
+[design]: #detailed-design
+
+### English
+[english]: #english
+
+This section applies to `rustc` and the standard library.
+
+An additional suggestion over RFC 505: One specific rule that comes up often:
+when quoting something for emphasis, use a single quote, and put punctuation
+outside the quotes, ‘this’. When quoting something at length, “use double
+quotes and put the punctuation inside of the quote.” Most documentation will
+end up using single quotes, so if you’re not sure, just stick with them.
+
+### Using Markdown
+[using-markdown]: #using-markdown
+
+The updated list of common headings is:
+
+* Examples
+* Panics
+* Errors
+* Safety
+* Aborts
+* Undefined Behavior
+
+RFC 505 suggests that one should always use the `rust` formatting directive:
+
+    ```rust
+    println!("Hello, world!");
+    ```
+
+    ```ruby
+    puts "Hello"
+    ```
+
+But, in API documentation, feel free to rely on the default being ‘rust’:
+
+    /// For example:
+    ///
+    /// ```
+    /// let x = 5;
+    /// ```
+
+Other places do not know how to highlight this anyway, so it's not important to
+be explicit.
+
+RFC 505 suggests that references and citation should be linked ‘reference
+style.’ This is still recommended, but prefer to leave off the second `[]`:
+
+```
+[Rust website]
+
+[Rust website]: http://www.rust-lang.org
+```
+
+to
+
+```
+[Rust website][website]
+
+[website]: http://www.rust-lang.org
+```
+
+But, if the text is very long, it is okay to use this form.
+
+### Examples in API docs
+[examples-in-api-docs]: #examples-in-api-docs
+
+Everything should have examples. Here is an example of how to do examples:
+
+```
+/// # Examples
+///
+/// Basic usage:
+///
+/// ```
+/// use op;
+///
+/// let s = "foo";
+/// let answer = op::compare(s, "bar");
+/// ```
+///
+/// Passing a closure to compare with, rather than a string:
+///
+/// ```
+/// use op;
+///
+/// let s = "foo";
+/// let answer = op::compare(s, |a| a.chars().is_whitespace().all());
+/// ```
+```
+
+For particularly simple APIs, still say “Examples” and “Basic usage:” for
+consistency’s sake.
+
+### Referring to types
+[referring-to-types]: #referring-to-types
+
+When talking about a type, use its full name. In other words, if the type is generic,
+say `Option<T>`, not `Option`. An exception to this is bounds. Write `Cow<'a, B>`
+rather than `Cow<'a, B> where B: 'a + ToOwned + ?Sized`.
+
+Another possibility is to write in lower case using a more generic term. In other words,
+‘string’ can refer to a `String` or an `&str`, and ‘an option’ can be ‘an `Option<T>`’.
+
+### Use parentheses for functions
+[use-parentheses-for-functions]: #use-parentheses-for-functions
+
+When referring to function names, include the `()`s after the name. For example, do this:
+
+```rust
+/// This behavior is similar to the way that `mem::replace()` works.
+```
```

Comment:
I'm not sure that this should be the suggested convention. This misleadingly suggests that [`mem::replace`] takes no arguments.

I think that the convention should either be to write out names for the arguments when referring to a function, such as `mem::replace(dest, src)`, or leave off the parentheses entirely, as in `mem::replace`.

The [JavaDoc Style Guide] addresses this, and makes the same recommendation; of course, Java has the additional complication of having overloaded methods, but I still think that even without them, it's still misleading to refer to a function or method with an empty parameter list when it takes parameters.

[`mem::replace`]: http://doc.rust-lang.org/std/mem/fn.replace.html
[JavaDoc Style Guide]: http://www.oracle.com/technetwork/articles/java/index-137868.html#styleguide